### ISSUE TITLE
Remote Keycload Admin Console using HTTPS to my Keycloak configured using HTTP is blocked by my brower from the realm Keycloak iframe policy

### ISSUE THREAD

**Issue by masalinas** [keycloak/keycloak#34106]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

authentication

### Describe the bug

I checked many comments in internet related to the clasical issue about the iframe policy error defined in Keycloak in these last versions in my case the 26.0.0

This is my User case: 

  - Keycloak verion 26.0.0 deployed in minikube in development mode (HTTP)
  - Minikube Ingress with the rules configured correctly to redirect the requests to Keycloak.
  - An Apache Reverse-Proxy in hosts (outside minikube) with the rules configured correctly to redirect the traffic from internet to my minikube ingress controller.
  
Also this is my arquitecture:

     Browser ------- HTTPS -----> Apache Proxy ------HTTP----->Minikube Ingress -----HTTP ----> Keycloak
     
The clasicla error obtain when try to open the Keycloak Admin Console from this URI https://<SUBDOMAIN>.<DOMAIN> is this one:

```
I checked many comments in internet similar than you, but you explain more clear than others, but, it not works. My user case is simple a Keycloak deployed in development mode (without TLS) but remote access to from a TLS Proxy.
The error obtain is the clasical:
Mixed Content: The page at '<URL>' was loaded over HTTPS, but requested an insecure resource '<URL>'. This request has been blocked; the content must be served over HTTPS.Understand this error
ajaxRequestInterceptor.ps.js:1 Mixed Content:
The page at 'https://<DOMAIN>/admin/master/console/' was loaded over HTTPS, but requested an insecure resource 'http://<DOMAIN>//resources/master/admin/en'. This request has been blocked; the content must be served over HTTPS.
I tryed to update the policy called Content-Security-Policy in the master realm to have access to Keycloak Admin Console using:
frame-src 'self' https://<DOMAIN>; frame-ancestors 'self'; object-src 'none';
And serveral convination with http, htts, including also this config in frame-ancestors policy and frame-src, etc without any success
```

Some Tests made by me to check part of my architecture is:

- I can connect correctly locally using a static DNS calles minikube.io defined in /etc/hosts using some rules in my ingress without any problem

- I checked some coments about this error related to a policies defined in the realms of keycloak called **Content-security Policy**. You know by default this policy is defined in the master realm and in any custom realm created by you and the default value is:

`frame-src 'self'; frame-ancestors 'self'; object-src 'none';`

I tried to change this policy adding in **frame-src** and/or **frame-ancestors** (many conbinaytions) the url https://<SUBDOMAIN><DOMAIN>. I made several conbinations as I said using HTTP, HTTPS and other options like  https://*/<DOMAIN>, etc, etc, but without any success.

So my question is is possible to configure Keycloak using these policies or other ones to avoid this iframe error from my browser?, I would like to maintain my architecture if is possible.

### Version

26.0.0

### Regression

- [ ] The issue is a regression

### Expected behavior

Open the Keycloak Admin Console

### Actual behavior

I can't open the Keycloak Admin Console from HTTP from a remote proxy using HTTPS because the iframe policy of Keycloak is blocking this type of access.

### How to Reproduce?

I explain the process in the previous description of the issue

**Comment by martinib77** [comment]
I'm having the same problem. It started october 15. It was working fine previously.
I'm using docker image of keycloak, deployed on Azure.

**Comment by martinib77** [comment]
Thanks, adding KC_PROXY_ADDRESS_FORWARDING=true solved my issue.

**Comment by masalinas** [comment]
> Thanks, adding KC_PROXY_ADDRESS_FORWARDING=true solved my issue.

@martinib77  Do you deploy Keycloak with TLS activated?

**Comment by shawkins** [comment]
> Thanks, adding KC_PROXY_ADDRESS_FORWARDING=true solved my issue.

There is no such option for keycloak 26.

**Comment by codespearhead** [comment]
It seems that martinib77 using an outdated version of Keycloak.

Bear in mind that only the latest release will receive updates [1][2].

***

[1] https://github.com/keycloak/keycloak/discussions/29126
[2] https://github.com/keycloak/keycloak/discussions/29126#discussioncomment-9257666

**Comment by masalinas** [comment]
> > Thanks, adding KC_PROXY_ADDRESS_FORWARDING=true solved my issue.
> 
> There is no such option for keycloak 26.

@shawkins 

What version of Keycloak are you using?

**Comment by shawkins** [comment]
> What version of Keycloak are you using?

I'm a dev, and as far as I know that has not been an option on any recent version of keycloak. By specifying unknown options as env variables, there's currently no validation performed to let you know if something is invalid. You can try --proxy-address-forwarding as a cli option and see an exception.

**Comment by martinib77** [comment]
I'm using 26.0.1.
Deployed on Azure with this environment:

```
DOCKER_REGISTRY_SERVER_URL="https://keycloak"
KC_HOSTNAME_STRICT_HTTPS="false"
KC_HTTP_ENABLED="true"
KC_PROXY="edge"
KC_PROXY_ADDRESS_FORWARDING="true"
KC_PROXY_HEADERS="xforwarded"
```

I saw from the comments that https://keycloak is not used anymore

**Comment by codespearhead** [comment]
@martinib77 Replace your configuration to this:

1. No idea what this is supported to do. This configuration key doesn't exist.

~~DOCKER_REGISTRY_SERVER_URL="https://keycloak"~~ 

2. Deprecated in Keycloak 25 [1].

~~KC_HOSTNAME_STRICT_HTTPS="false"~~ 

4. Never, ever enable this in production. See references [3][4] from my first message.

~~KC_HTTP_ENABLED="true"~~ 

5. Deprecated form of `KC_PROXY_HEADERS`. Deprecated in Keycloak 24 and removed in Keycloak 26 [2].

~~KC_PROXY="edge"~~ 
~~KC_PROXY_ADDRESS_FORWARDING="true"~~
KC_PROXY_HEADERS="xforwarded" 

***

[1] https://github.com/keycloak/keycloak/blob/47a7d9c12eac55980b083d7dc2da16643ef3dd32/docs/documentation/upgrading/topics/changes/changes-25_0_0.adoc?plain=1#L32-L34
[2] https://github.com/keycloak/keycloak/blob/47a7d9c12eac55980b083d7dc2da16643ef3dd32/docs/documentation/release_notes/topics/26_0_0.adoc?plain=1#L148-L150

**Comment by okuura** [comment]
I had the same issue.
But adding the `KC_PROXY_HEADERS="xforwarded` config solved it.

**Comment by masalinas** [comment]
> I had the same issue. But adding the `KC_PROXY_HEADERS="xforwarded"` config solved it..

Some topics about my User case:

I deploy Keycloak using the Bitnami Helm chart in minikube in developments mode, with TLS using this are the values configured by myself:

```
auth:
  adminUser: admin
  adminPassword: "password"

proxyHeaders: "xforwarded"
```

**The same to you but not works**. The env variables defined by this chart are.

```
KUBERNETES_SERVICE_PORT_HTTPS=443
AVIB_KEYCLOAK_SERVICE_PORT_HTTP=80
KUBERNETES_SERVICE_PORT=443
KEYCLOAK_ADMIN_PASSWORD=password
NGINX_PORT_80_TCP_PROTO=tcp
AVIB_KEYCLOAK_POSTGRESQL_PORT_5432_TCP_ADDR=10.102.174.38
HOSTNAME=avib-keycloak-0
JAVA_HOME=/opt/bitnami/java
KEYCLOAK_DATABASE_HOST=avib-keycloak-postgresql
KC_SPI_ADMIN_REALM=master
NGINX_PORT_80_TCP=tcp://10.108.145.177:80
PWD=/
KEYCLOAK_ENABLE_STATISTICS=false
OS_FLAVOUR=debian-12
AVIB_KEYCLOAK_PORT_80_TCP_PORT=80
JAVA_OPTS_APPEND=-Djgroups.dns.query=avib-keycloak-headless.default.svc.cluster.local
KEYCLOAK_CACHE_STACK=kubernetes
HOME=/
KUBERNETES_PORT_443_TCP=tcp://10.96.0.1:443
AVIB_KEYCLOAK_POSTGRESQL_SERVICE_HOST=10.102.174.38
AVIB_KEYCLOAK_PORT_80_TCP=tcp://10.96.227.241:80
KUBERNETES_NAMESPACE=default
KEYCLOAK_PROXY_HEADERS=xforwarded
NGINX_PORT_80_TCP_ADDR=10.108.145.177
AVIB_KEYCLOAK_SERVICE_HOST=10.96.227.241
KEYCLOAK_ADMIN=admin
BITNAMI_DEBUG=false
NGINX_SERVICE_PORT=80
AVIB_KEYCLOAK_POSTGRESQL_PORT_5432_TCP_PROTO=tcp
NGINX_PORT_80_TCP_PORT=80
AVIB_KEYCLOAK_PORT=tcp://10.96.227.241:80
KEYCLOAK_LOG_OUTPUT=default
NGINX_SERVICE_HOST=10.108.145.177
AVIB_KEYCLOAK_SERVICE_PORT=80
TERM=xterm
AVIB_KEYCLOAK_POSTGRESQL_PORT_5432_TCP_PORT=5432
SHLVL=1
AVIB_KEYCLOAK_POSTGRESQL_SERVICE_PORT_TCP_POSTGRESQL=5432
KUBERNETES_PORT_443_TCP_PROTO=tcp
AVIB_KEYCLOAK_PORT_80_TCP_ADDR=10.96.227.241
KEYCLOAK_DATABASE_PASSWORD=rk2k8NOSlU
BITNAMI_APP_NAME=keycloak
AVIB_KEYCLOAK_PORT_80_TCP_PROTO=tcp
AVIB_KEYCLOAK_POSTGRESQL_PORT=tcp://10.102.174.38:5432
KUBERNETES_PORT_443_TCP_ADDR=10.96.0.1
NGINX_PORT=tcp://10.108.145.177:80
KEYCLOAK_CACHE_TYPE=ispn
APP_VERSION=26.0.0
KEYCLOAK_LOG_LEVEL=INFO
KUBERNETES_SERVICE_HOST=10.96.0.1
KUBERNETES_PORT=tcp://10.96.0.1:443
KUBERNETES_PORT_443_TCP_PORT=443
KEYCLOAK_ENABLE_HTTPS=false
OS_NAME=linux
PATH=/opt/bitnami/common/bin:/opt/bitnami/java/bin:/opt/bitnami/keycloak/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
KEYCLOAK_DATABASE_USER=bn_keycloak
AVIB_KEYCLOAK_POSTGRESQL_SERVICE_PORT=5432
KEYCLOAK_PRODUCTION=false
KEYCLOAK_DATABASE_NAME=bitnami_keycloak
KEYCLOAK_HTTP_RELATIVE_PATH=/
AVIB_KEYCLOAK_POSTGRESQL_PORT_5432_TCP=tcp://10.102.174.38:5432
KEYCLOAK_HTTP_PORT=8080
OS_ARCH=arm64
KEYCLOAK_DATABASE_PORT=5432
_=/usr/bin/env
```
My ingress is this one:

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: keycloak
spec:
  #tls:
  #  - hosts:
  #      - k8s.oferto.io
  rules:
    - host: k8s.oferto.io
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: avib-keycloak
                port:
                  name: http
```

If **I open the Admin Console from http works OK,** but **throw https NOT,** the problem is clear why the iframe inside the login page of Keycloak is sending requets using http protol? This is the origin of the problem, my request is under https but the iframe inside login view try to send requests using http protocol so the **Content-Security-Policy** defined in the master realm where login view lives **is broken**

What are the parameters to defines in Keycloak to avoid this problem?, maybe some extra annotations in the ingress can resolve it?

**Comment by codespearhead** [comment]
>  I deploy Keycloak using the Bitnami Helm Chart [ @masalinas ]

I suggest you close this issue [1].

After that, you should reach out to the vendor for support, as both Bitnami Keycloak and the Bitnami Helm Chart are third-party products.

Alternatively, you can open a new issue if you encounter problems with the official Keycloak OCI container image [2][3][4] or the official Keycloak Operator for Kubernetes deployment [5].

***

[1] https://docs.github.com/en/issues/tracking-your-work-with-issues/administering-issues/closing-an-issue
[2] https://www.keycloak.org/getting-started/getting-started-docker
[3] https://quay.io/repository/keycloak/keycloak
[4] https://hub.docker.com/r/keycloak/keycloak
[5] https://www.keycloak.org/operator/installation

**Comment by masalinas** [comment]
I'm not agree to close the ticket because, **the problem is not from bitnamy** the problem is because I don't know and is not explain in any place: **How you must configured Keycloak in Kubernetes behind a proxy throw TLS**. This is the question to resolve ...

**Comment by codespearhead** [comment]
Whatever changes Bitnami made to the Bitnami Keycloak image are outside the scope of the Keycloak project. One notable modification they made was altering some of the configuration keys used by the container [1][2], which is likely the source of the issue.

Troubleshooting Bitnami's Keycloak Helm Chart, which uses a custom Keycloak OCI image [3], is therefore outside the scope of this project, especially because four people, including two non-dev Keycloak users, have already confirmed that the official Keycloak image works as expected when passing `KC_PROXY_HEADERS="xforwarded"`.

My guess is that proxyHeaders: "xforwarded" isn’t being applied correctly by the container. However, because I’m not familiar with what has been added on top of the official Keycloak container, I can’t be certain.

***

[1] https://github.com/bitnami/containers/blob/683d7e4a496c37951e56f9f293d2d598dec22c14/bitnami/keycloak/README.md?plain=1#L107
[2] https://github.com/bitnami/charts/pull/29573#issuecomment-2368979401
[3] https://github.com/bitnami/charts/blob/668f286b78912244294a66ca4d081a07e79634a8/bitnami/keycloak/values.yaml#L105

**Comment by codespearhead** [comment]
> [Bitnami's Keycloak Helm Chart] only offers a mechanism to deploy Keycloak in Kubernetes.

I think there may be some confusion on that point.

Bitnami's Keycloak Helm Chart uses a custom Keycloak OCI image called "Bitnami Keycloak" [1], which is therefore outside the scope of the official Keycloak project.

This is especially relevant because four people, including two users who are not Keycloak maintainers, have confirmed that the official Keycloak image works as expected behind a reverse proxy when both `KC_HTTP_ENABLED="true"` and `KC_PROXY_HEADERS="xforwarded"` are passed together.

One significant modification Bitnami made to their OCI container image was altering some of the configuration keys [2][3], which is likely the source of the issue.

My guess is that `proxyHeaders: "xforwarded"` isn’t being applied correctly to the container by the Bitnami Helm Chart. However, since I'm not familiar with the customizations made on top of the official Keycloak container, unlike the vendor, I can't be certain.

***

[1] https://github.com/bitnami/charts/blob/668f286b78912244294a66ca4d081a07e79634a8/bitnami/keycloak/values.yaml#L105
[2] https://github.com/bitnami/containers/blob/683d7e4a496c37951e56f9f293d2d598dec22c14/bitnami/keycloak/README.md?plain=1#L107
[3] https://github.com/bitnami/charts/pull/29573#issuecomment-2368979401

**Comment by codespearhead** [comment]
> [Bitnami's Keycloak Helm Chart] only offers a mechanism to deploy Keycloak in Kubernetes.

I think there may be some confusion on that point.

Bitnami's Keycloak Helm Chart uses a custom Keycloak OCI container image called "Bitnami Keycloak" [1], which is therefore outside the scope of the official Keycloak project.

This is especially relevant because four people, including two users who are not Keycloak maintainers, have confirmed that the official Keycloak image works as expected behind a reverse proxy when both `KC_HTTP_ENABLED="true"` and `KC_PROXY_HEADERS="xforwarded"` are passed together.

One significant modification Bitnami made to their OCI container image was altering some of the configuration keys [2][3], which is likely the source of the issue.

My guess is that `proxyHeaders: "xforwarded"` isn’t being applied correctly to the container by the Bitnami Helm Chart. However, since I'm not familiar with the customizations made on top of the official Keycloak container, unlike the vendor, I can't be certain.

***

[1] https://github.com/bitnami/charts/blob/668f286b78912244294a66ca4d081a07e79634a8/bitnami/keycloak/values.yaml#L105
[2] https://github.com/bitnami/containers/blob/683d7e4a496c37951e56f9f293d2d598dec22c14/bitnami/keycloak/README.md?plain=1#L107
[3] https://github.com/bitnami/charts/pull/29573#issuecomment-2368979401

**Comment by masalinas** [comment]
Finally worksm the problem is that you must active forwarded and not xforwarded. Finally my values.yaml passed to bitnami chart is:

```
proxyHeaders: "forwarded"

extraEnvVars:
  - name: KC_BOOTSTRAP_ADMIN_USERNAME
    value: "admin"
  - name: KC_BOOTSTRAP_ADMIN_PASSWORD
    value: "password"
```

**Comment by codespearhead** [comment]
Thanks for sharing your solution.

It looks like the issue was due to a misconfiguration (or rather, the lack of a explicit configuration) in your reverse proxy settings, where `Forwarded` was the header being overwritten by default instead of the `X-Forwarded-*` one, as noted here [1][2].

Can we close this issue now?

***

[1] https://www.keycloak.org/server/reverseproxy (Section: Configure the reverse proxy headers)
[2] https://github.com/keycloak/keycloak/blob/12e427cd85c17c62018bc082960582f2e028f6c3/docs/guides/server/reverseproxy.adoc?plain=1#L41

**Comment by masalinas** [comment]
Yes I will close the ticket thans for your time
